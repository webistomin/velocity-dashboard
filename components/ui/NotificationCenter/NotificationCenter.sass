+b('notification-center')
  +e('btn')
    width: rem(50px)
    height: rem(50px)
    max-width: rem(60px)
    display: flex
    align-items: center
    justify-content: center
    color: var(--color-gray-4)
    transition: color 0.2s ease-out

    &:hover
      color: var(--color-primary)

  +e('btn-content')
    position: relative

    .icon
      width: rem(20px)
      height: rem(22px)

    +media($screen-lg)
      .icon
        width: rem(26px)
        height: rem(24px)

  +e('sidebar')
    width: 282px
    height: calc(100% - 5.1rem)
    top: rem(51px)
    right: 0

    +media($screen-sm)
      top: 8rem
      height: calc(100% - 8rem)

  +e('overlay')
    height: calc(100% - 5.1rem)
    top: rem(51px)

    +media($screen-sm)
      top: 8rem
      height: calc(100% - 8rem)

    +media($screen-lg)
      display: none

  +e('item')
    border-bottom: rem(1px) solid var(--color-border-gray)

  +e('item-link')
    display: flex
    box-sizing: border-box
    width: 100%
    height: 100%
    padding: rem(25px) rem(14px)

    +media($screen-sm)
      padding-left: rem(15px)
      padding-right: rem(15px)

  +e('col')
    +m('left')
      position: relative
      margin-right: rem(15px)
      margin-top: rem(28px)

  +e('item-icon')
    color: var(--color-gray-3)

    +m('new')
      color: var(--color-primary)

  +e('caption')
    margin-bottom: rem(10px)

  +e('text')
    margin-bottom: rem(7px)
    +line-clamp(2)

  +e('time')
    font: var(--text-caption)
    font-weight: 400
    color: var(--color-gray-4)

  +e('heading')
    display: flex
    justify-content: space-between
    align-items: center
    padding: rem(25px) rem(14px)
    border-bottom: rem(1px) solid var(--color-border-gray)

    +media($screen-sm)
      padding-left: rem(24px)
      padding-right: rem(24px)

    +media($screen-lg)
      padding-top: rem(30px)
      padding-bottom: rem(30px)

  +e('clear-btn')
    width: rem(90px)

  +e('city')
    color: var(--color-gray-3)
    font-weight: 400

  +e('weather-icon')
    margin-left: rem(14px)
