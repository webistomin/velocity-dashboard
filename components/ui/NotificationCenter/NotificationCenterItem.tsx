import { VNode } from 'vue';
import { BaseIcon } from 'components/base/BaseIcon/BaseIcon';
import { VueComponent } from 'types/vue-components';
import { BaseNotify } from 'components/base/BaseNotify/BaseNotify';
import { Component, Prop } from '~/node_modules/nuxt-property-decorator';

export enum NotificationCenterItemTypes {
  MESSAGE = 'message',
  APPLICATION = 'application',
  COMPLAINT = 'complaint',
}

export interface INotificationCenterItemProps {
  id: string | number;
  type: NotificationCenterItemTypes;
  text: string;
  time: DOMTimeStamp;
  isNew: boolean;
}

@Component({
  name: 'NotificationCenterItem',
})
export default class NotificationCenterItem extends VueComponent<INotificationCenterItemProps> {
  @Prop({ default: NotificationCenterItemTypes.APPLICATION })
  private readonly type!: INotificationCenterItemProps['type'];

  @Prop({ default: '' })
  private readonly text!: INotificationCenterItemProps['text'];

  @Prop({ default: '' })
  private readonly time!: INotificationCenterItemProps['time'];

  @Prop({ default: '' })
  private readonly id!: INotificationCenterItemProps['id'];

  @Prop({ default: false })
  private readonly isNew!: INotificationCenterItemProps['isNew'];

  get getIconName() {
    switch (this.type) {
      case NotificationCenterItemTypes.MESSAGE:
        return 'icon-bubble';
      case NotificationCenterItemTypes.APPLICATION:
        return 'icon-mail';
      case NotificationCenterItemTypes.COMPLAINT:
        return 'icon-lightning';
      default:
        return 'icon-mail';
    }
  }

  render(): VNode {
    return (
      <div class='notification-center__item'>
        <nuxt-link to='/test' class='notification-center__item-link link'>
          <div class='notification-center__col notification-center__col_left'>
            {this.isNew ? <BaseNotify/> : null}
            <BaseIcon
              class={`notification-center__item-icon ${this.isNew ? 'notification-center__item-icon_new' : ''}`}
              name={this.getIconName}
              size='xs'
            />
          </div>
          <div class='notification-center__col notification-center__col_right'>
            <strong class='notification-center__caption caption'>{this.type.toUpperCase()}</strong>
            <p class='notification-center__text paragraph'>{this.text}</p>
            <time class='notification-center__time'>{this.time}</time>
          </div>
        </nuxt-link>
      </div>
    );
  }
}
