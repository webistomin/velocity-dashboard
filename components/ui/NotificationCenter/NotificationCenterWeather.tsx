import { VueComponent } from 'types/vue-components';
import { Component, Prop } from 'nuxt-property-decorator';
import { VNode } from 'vue';

import BaseTitle from 'components/base/BaseTitle';
import { ICurrentWeather } from 'common/types/weather/current';

export interface INotificationCenterWeatherProps {
  city: string;
  temperature: ICurrentWeather['temp'];
  icon: ICurrentWeather['weather']['icon'];
  description: ICurrentWeather['weather']['description'];
}

@Component({
  name: 'NotificationCenterWeather',
})
export default class NotificationCenterWeather extends VueComponent<INotificationCenterWeatherProps> {
  @Prop({ default: '' })
  private readonly city!: INotificationCenterWeatherProps['city'];

  @Prop({ default: 0 })
  private readonly temperature!: INotificationCenterWeatherProps['temperature'];

  @Prop({ default: '' })
  private readonly icon!: INotificationCenterWeatherProps['icon'];

  @Prop({ default: '' })
  private readonly description!: INotificationCenterWeatherProps['description'];

  public render(): VNode {
    return (
      <div class='notification-center__weather' title={this.description}>
        <BaseTitle level={3} class='notification-center__weather-title'>
          <span>{this.temperature}ÂºF</span>
          <v-lazy-image
            src={`/img/weather-icons/${this.icon}.png`}
            alt={this.description}
            width={48}
            height={48}
            class='notification-center__weather-icon image'
          />
        </BaseTitle>
        <strong class='notification-center__city'>{this.city}</strong>
      </div>
    );
  }
}
