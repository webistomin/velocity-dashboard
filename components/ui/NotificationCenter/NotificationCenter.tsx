import { VueComponent } from 'types/vue-components';
import { Component, Watch } from 'nuxt-property-decorator';
import { VNode } from 'vue';
import { Getter, Action, State } from 'vuex-class';

import BaseNotify from 'components/base/BaseNotify';
import BaseOverlay from 'components/base/BaseOverlay';
import BaseSidebar from 'components/base/BaseSidebar';
import BaseButton from 'components/base/BaseButton';
import { IGetCurrentWeatherResult, IWeatherCurrentCoordinates } from 'common/types/weather/current';
import { BaseSpinner } from 'components/base/BaseSpinner/BaseSpinner';
import NotificationCenterItem, {
  INotificationCenterItemProps,
  NotificationCenterItemTypes,
} from './NotificationCenterItem';
import { NotificationCenterWeather } from './NotificationCenterWeather';

import './NotificationCenter.sass';

@Component({
  name: 'NotificationCenter',
})
export default class NotificationCenter extends VueComponent {
  public isNotificationsVisible: boolean = false;

  public notifications: INotificationCenterItemProps[] = [
    {
      type: NotificationCenterItemTypes.MESSAGE,
      text: 'How may we assist you today?',
      time: 1591086090152,
      id: '1',
      isNew: true,
    },
    {
      type: NotificationCenterItemTypes.APPLICATION,
      text: 'How may we assist you today?',
      time: 1591086090152,
      id: '2',
      isNew: false,
    },
    {
      type: NotificationCenterItemTypes.COMPLAINT,
      text: 'How may we assist you today?',
      time: 1591086090152,
      id: '3',
      isNew: true,
    },
    {
      type: NotificationCenterItemTypes.MESSAGE,
      text: 'How may we assist you today?',
      time: 1591086090152,
      id: '4',
      isNew: true,
    },
    {
      type: NotificationCenterItemTypes.APPLICATION,
      text: 'How may we assist you today?',
      time: 1591086090152,
      id: '5',
      isNew: true,
    },
    {
      type: NotificationCenterItemTypes.APPLICATION,
      text: 'How may we assist you today?',
      time: 1591086090152,
      id: '6',
      isNew: false,
    },
    {
      type: NotificationCenterItemTypes.MESSAGE,
      text: 'How may we assist you today?',
      time: 1591086090152,
      id: '7',
      isNew: true,
    },
    {
      type: NotificationCenterItemTypes.COMPLAINT,
      text: 'How may we assist you today?',
      time: 1591086090152,
      id: '8',
      isNew: false,
    },
  ];

  @Action('weather/fetchCurrentWeather') private readonly fetchCurrentWeather!: () => void;

  @Getter('weather/getCurrentWeather') getCurrentWeather!: IGetCurrentWeatherResult;

  @State((state) => state.weather.coordinates)
  private readonly getUserCoordinates!: IWeatherCurrentCoordinates;

  @Watch('$route')
  public onRouteChanged() {
    this.isNotificationsVisible = false;
  }

  public toggleNotificationVisibility() {
    if (!this.isNotificationsVisible && !this.getCurrentWeather) {
      this.fetchCurrentWeather();
    }

    this.isNotificationsVisible = !this.isNotificationsVisible;
  }

  public get shouldLockBody(): boolean {
    return this.isNotificationsVisible && window.matchMedia('(max-width: 1023px)').matches;
  }

  public onClickOutside(): void {
    this.isNotificationsVisible = false;
  }

  public toggleModalVisibilityByKeyDown(event: KeyboardEvent) {
    if (this.isNotificationsVisible && event.key === 'Escape') {
      this.onClickOutside();
    }
  }

  public mounted(): void {
    document.addEventListener('keydown', this.toggleModalVisibilityByKeyDown);
  }

  public beforeDestroy(): void {
    document.removeEventListener('keydown', this.toggleModalVisibilityByKeyDown);
  }

  public render(): VNode {
    return (
      <div class='notification-center' v-scroll-lock={this.shouldLockBody} v-click-outside={this.onClickOutside}>
        <button
          class={`notification-center__btn btn ${this.isNotificationsVisible ? 'notification-center__btn_active' : ''}`}
          type='button'
          onClick={this.toggleNotificationVisibility}
          aria-label='Toggle notification center'>
          <span class='notification-center__btn-content'>
            <BaseNotify class='notification-center__status' />
            <svg-icon name='icon-bell' width={20} height={22} />
          </span>
        </button>

        <BaseSidebar isVisible={this.isNotificationsVisible} class='notification-center__sidebar'>
          <div class='notification-center__heading'>
            <div class='notification-center__info'>
              {this.getCurrentWeather ? (
                <NotificationCenterWeather
                  temperature={this.getCurrentWeather.temperature}
                  city={`${this.getCurrentWeather.cityName}, ${this.getCurrentWeather.countryCode}`}
                  icon={this.getCurrentWeather.icon}
                  description={this.getCurrentWeather.description}
                />
              ) : this.getUserCoordinates ? (
                <BaseSpinner size='s' />
              ) : null}
            </div>
            <BaseButton theme='gray' class='notification-center__clear-btn'>
              Clear
            </BaseButton>
          </div>
          <ul class='notification-center__list list'>
            {this.notifications.map((notification) => {
              return (
                <li class='notification-center__list-item list-item' key={notification.id}>
                  <NotificationCenterItem
                    type={notification.type}
                    text={notification.text}
                    time={notification.time}
                    id={notification.id}
                    isNew={notification.isNew}
                  />
                </li>
              );
            })}
          </ul>
        </BaseSidebar>

        <BaseOverlay
          class='notification-center__overlay'
          isVisible={this.isNotificationsVisible}
          onClick={this.onClickOutside}
        />
      </div>
    );
  }
}
