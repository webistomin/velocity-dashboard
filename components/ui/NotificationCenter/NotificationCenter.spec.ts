import { shallowMount, createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
// @ts-ignore
import vClickOutside from 'v-click-outside';
import VScrollLock from 'v-scroll-lock';
import NotificationCenter from './NotificationCenter';

describe('NotificationCenter UI', () => {
  const localVue = createLocalVue();
  localVue.use(VueRouter);
  localVue.use(vClickOutside);
  localVue.use(VScrollLock);
  localVue.use(Vuex);
  let store: any;
  let state;
  let getters;
  let actions;
  const router = new VueRouter();

  beforeEach(() => {
    state = {
      coordinates: {
        lat: 1,
        lng: 1,
      },
    };

    getters = {
      getCurrentWeather: jest.fn(() => {
        return {
          cityName: '',
          countryCode: '',
          description: '',
          icon: '',
          temperature: 1,
        };
      }),
    };

    actions = {
      fetchCurrentWeather: jest.fn().mockImplementation(() => Promise.resolve('called')),
    };

    store = new Vuex.Store({
      modules: {
        weather: {
          namespaced: true,
          state,
          getters,
          actions,
        },
      },
    });
  });

  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      value: jest.fn(() => {
        return {
          matches: true,
          addEventListener: jest.fn(),
          removeEventListener: jest.fn(),
        };
      }),
      writable: true,
    });
  });

  const factory = (propsData = {}, options = {}) => {
    return shallowMount(NotificationCenter, {
      localVue,
      store,
      router,
      propsData: {
        ...propsData,
      },
      stubs: ['svg-icon'],
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Toggle notifications visibility', () => {
    const wrapper = factory();
    wrapper.setData({
      isNotificationsVisible: false,
    });
    wrapper.vm.toggleNotificationVisibility();
    expect(wrapper.vm.isNotificationsVisible).toEqual(true);
  });

  it('Close modal on "Esc"', () => {
    const mockedCloseModal = jest.fn();
    const wrapper = factory(
      {},
      {
        attachToDocument: true,
        methods: {
          toggleModalVisibilityByKeyDown: mockedCloseModal,
        },
      }
    );
    wrapper.trigger('keydown.esc', { key: 'Escape' });
    expect(mockedCloseModal).toHaveBeenCalled();
  });

  it('Delete global event listeners on destroy', () => {
    const mockedCloseModal = jest.fn();
    const wrapper = factory(
      {},
      {
        attachToDocument: true,
        methods: {
          toggleModalVisibilityByKeyDown: mockedCloseModal,
        },
      }
    );
    wrapper.destroy();
    wrapper.trigger('keydown.esc', { key: 'Escape' });
    expect(mockedCloseModal).not.toHaveBeenCalled();
  });

  it('Close modal on route change', async () => {
    const wrapper = factory();
    wrapper.setData({
      isNotificationsVisible: true,
    });
    await wrapper.vm.$router.push({ hash: '#' });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.isNotificationsVisible).toEqual(false);
  });
});
