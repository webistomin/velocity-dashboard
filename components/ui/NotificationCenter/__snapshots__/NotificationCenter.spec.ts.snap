// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationCenter UI Matches snapshot 1`] = `
"<div class=\\"notification-center\\"><button type=\\"button\\" aria-label=\\"Toggle notification center\\" class=\\"notification-center__btn btn \\"><span class=\\"notification-center__btn-content\\"><vuecomponent-stub class=\\"notification-center__status\\"></vuecomponent-stub><svg-icon-stub name=\\"icon-bell\\" width=\\"20\\" height=\\"22\\"></svg-icon-stub></span></button>
  <vuecomponent-stub class=\\"notification-center__sidebar\\">
    <div class=\\"notification-center__heading\\">
      <div class=\\"notification-center__info\\">
        <vuecomponent-stub city=\\", \\" temperature=\\"1\\" icon=\\"\\" description=\\"\\"></vuecomponent-stub>
      </div>
      <basebutton-stub type=\\"button\\" icon=\\"\\" theme=\\"gray\\" class=\\"notification-center__clear-btn\\">Clear</basebutton-stub>
    </div>
    <ul class=\\"notification-center__list list\\">
      <li class=\\"notification-center__list-item list-item\\">
        <notificationcenteritem-stub type=\\"message\\" text=\\"How may we assist you today?\\" time=\\"1591086090152\\" id=\\"1\\" isnew=\\"true\\"></notificationcenteritem-stub>
      </li>
      <li class=\\"notification-center__list-item list-item\\">
        <notificationcenteritem-stub type=\\"application\\" text=\\"How may we assist you today?\\" time=\\"1591086090152\\" id=\\"2\\"></notificationcenteritem-stub>
      </li>
      <li class=\\"notification-center__list-item list-item\\">
        <notificationcenteritem-stub type=\\"complaint\\" text=\\"How may we assist you today?\\" time=\\"1591086090152\\" id=\\"3\\" isnew=\\"true\\"></notificationcenteritem-stub>
      </li>
      <li class=\\"notification-center__list-item list-item\\">
        <notificationcenteritem-stub type=\\"message\\" text=\\"How may we assist you today?\\" time=\\"1591086090152\\" id=\\"4\\" isnew=\\"true\\"></notificationcenteritem-stub>
      </li>
      <li class=\\"notification-center__list-item list-item\\">
        <notificationcenteritem-stub type=\\"application\\" text=\\"How may we assist you today?\\" time=\\"1591086090152\\" id=\\"5\\" isnew=\\"true\\"></notificationcenteritem-stub>
      </li>
      <li class=\\"notification-center__list-item list-item\\">
        <notificationcenteritem-stub type=\\"application\\" text=\\"How may we assist you today?\\" time=\\"1591086090152\\" id=\\"6\\"></notificationcenteritem-stub>
      </li>
      <li class=\\"notification-center__list-item list-item\\">
        <notificationcenteritem-stub type=\\"message\\" text=\\"How may we assist you today?\\" time=\\"1591086090152\\" id=\\"7\\" isnew=\\"true\\"></notificationcenteritem-stub>
      </li>
      <li class=\\"notification-center__list-item list-item\\">
        <notificationcenteritem-stub type=\\"complaint\\" text=\\"How may we assist you today?\\" time=\\"1591086090152\\" id=\\"8\\"></notificationcenteritem-stub>
      </li>
    </ul>
  </vuecomponent-stub>
  <baseoverlay-stub class=\\"notification-center__overlay\\"></baseoverlay-stub>
</div>"
`;
